### ISSUE TITLE
Best VPN protocol setup for a VPS

### ISSUE THREAD

**Issue by Msadr471** [net4people/bbs#180]
Hi,
I am from Iran and recently bought a VPS on Hetzner, last night tried to set up a wireguard but it was so confusing that I used this [source](https://www.digitalocean.com/community/tutorials/how-to-set-up-wireguard-on-ubuntu-20-04)  to set it up and it did not work I mean I know I did something wrong but don't know in which part! the most confusing part was that every site had its own different setup! so the question is wireguard good? do you have any sources that help me? or should I choose a different protocol cause I'm living in Iran and wireguard won't work? 
I'm not an IT or network Engineer, for example:
When I was setting up wireguard last night:

```
[Interface]
PrivateKey = base64_encoded_peer_private_key_goes_here
Address = 10.8.0.2/24
Address = fd0d:86fa:c3bc::2/64

[Peer]
PublicKey = ****************************************
AllowedIPs = 10.8.0.0/24, fd0d:86fa:c3bc::/64
Endpoint = 203.0.113.1:51820
```

I didn't know what I should choose for my `Address` or `Endpoint` or `AllowedIPs` OR in the server config itself! which IP?
I've seen this [repo](https://github.com/Nyr/openvpn-install) too and I get nothing, does OpenVPN work in Iran?
or does the government only block those servers that were providing VPNs like Nord or ProtonVPN? if I have access to my server so any kind of setup must work, right??? I don't know you tell me.
[here](https://github.com/net4people/bbs/issues/174) @arandomgstring says:

> There exists many one-click scripts that setup v2ray on server without need of python, https://github.com/reeceyng/v2ray-agent , https://github.com/wulabing/Xray_onekey , https://github.com/proxysu/ProxySU to name a few. 

so **WHAT THE HELL IS THIS**?
I barely understand and write English!

```
注意事项
修改Cloudflare->SSL/TLS->Overview->Full
Cloudflare ---> A记录解析的云朵必须为灰色【如非灰色，会影响到定时任务自动续签证书】
如用CDN又同时使用直连，关闭云朵+自选IP，自选IP参考上方的[Cloudflare 优化方案](https://github.com/reeceyng/v2ray-agent/blob/master/documents/optimize_V2Ray.md)
使用纯净系统安装，如使用其他脚本安装过并且自己无法修改错误，请重新安装系统后再次尝试安装
wget: command not found [这里需要自己手动安装下wget] ，如未使用过Linux，[点击查看](https://github.com/reeceyng/v2ray-agent/tree/master/documents/install_tools.md)安装教程
不支持非root账户
如发现Nginx相关问题，请卸载掉自编译的nginx或者重新安装系统
为了节约时间，反馈请带上详细截图或者按照模版规范，无截图或者不按照规范的issue会被直接关闭
不推荐GCP用户使用
不推荐使用Centos以及低版本的系统，如果Centos安装失败，请切换至Debian10重新尝试，脚本不再支持Centos6、Ubuntu 16.x
[如有使用不明白的地方请先查看脚本使用指南](https://github.com/reeceyng/v2ray-agent/blob/master/documents/how_to_use.md)
Oracle Cloud有一个额外的防火墙，需要手动设置
Oracle Cloud仅支持Ubuntu
如果使用gRPC通过cloudflare转发,需要在cloudflare设置允许gRPC，路径：cloudflare Network->gRPC
gRPC目前处于测试阶段，可能对你使用的客户端不兼容，如不能使用请忽略
低版本脚本升级高版本时无法启动问题，[请点击此链接查看解决方案](https://github.com/reeceyng/v2ray-agent/blob/master/documents/how_to_use.md#4%E4%BD%8E%E7%89%88%E6%9C%AC%E5%8D%87%E7%BA%A7%E9%AB%98%E7%89%88%E6%9C%AC%E5%90%8E%E6%97%A0%E6%B3%95%E5%90%AF%E5%8A%A8%E6%A0%B8%E5%BF%83)
```
BUT found [this](https://github.com/HirbodBehnam/V2Ray-Installer)  and also @arandomgstring says:

> and I am pretty sure it's not the only one.

so I'm all ears, is there any script for me that knows nothing to build or set up a VPN? easy script for setting up VPNs on my VPS.

thanks and sorry for my English.

**Comment by cross-hello** [comment]
Maybe the following repository could one bash line deploy:
https://github.com/maplecool/easytrojan

**Comment by free-the-internet** [comment]
> tell us if you find out! Xray+Vless+TCP+TLS seem to be the most resilient. but it doesn't protect against the active probe (see [#166 (comment)](https://github.com/net4people/bbs/issues/166#issuecomment-1356813450)). you need to have a camouflage website on port 443 with HTTPS and redirect suspicious packets to this endpoint so the censor does not add your domain/IP to _the list_.

Do you have a tutorial for having fallback website for Vless + TCP + TLS? I see one here: https://henrywithu.com/coexistence-of-web-applications-and-vless-tcp-xtls/ but it looks complex, I feel it must be simpler than that.
FYI, I used this: https://github.com/v2fly/v2ray-examples/tree/master/VLESS-TCP-TLS-proxy%20protocol, but I get nginx config error.

**Comment by bensafai** [comment]
> > tell us if you find out! Xray+Vless+TCP+TLS seem to be the most resilient. but it doesn't protect against the active probe (see [#166 (comment)](https://github.com/net4people/bbs/issues/166#issuecomment-1356813450)). you need to have a camouflage website on port 443 with HTTPS and redirect suspicious packets to this endpoint so the censor does not add your domain/IP to _the list_.
> 
> Do you have a tutorial for having fallback website for Vless + TCP + TLS? I see one here: https://henrywithu.com/coexistence-of-web-applications-and-vless-tcp-xtls/ but it looks complex, I feel it must be simpler than that. FYI, I used this: https://github.com/v2fly/v2ray-examples/tree/master/VLESS-TCP-TLS-proxy%20protocol, but I get nginx config error.

see this repo: https://github.com/reeceyng/v2ray-agent
It's work for me.

**Comment by bensafai** [comment]
> 

Yea, but unfortunately, more information about mack-a is not available.

**Comment by free-the-internet** [comment]
> > Do you have a tutorial for having fallback website for Vless + TCP + TLS?
> 
> no, I think @arandomgstring has set it up like this. I think he has used a derivation of this: https://github.com/XTLS/Xray-examples/tree/main/VLESS-TCP-TLS-WS%20(recommended)

@arandomgstring, could you please share your setup that has Vless + TCP + TLS + fallback to a mock site (nginx)?

BTW, thanks @Azadzadeh and @bensafai

**Comment by arandomgstring** [comment]
@pirooz-gthb 
That's not right. You have probably seen their english translated website which is 1 year behind the actual development of v2ray. See it yourself, https://www.v2fly.org/config/protocols/vless.html, do you see anything related to deprecation? Absolutely not. And even RPRX him/herself on a topic (I can't find it now) said that VLESS can be considered more secure than Trojan, because Trojan produces a "particular" socks5 like traffic underhood, and I can't agree more.



@Msadr471 

You can always use google translator, Deepl for their whole bash file, or you can translate line by line, you know... I am saying this, because original Chinese resources are always far ahead.

**Comment by arandomgstring** [comment]
@Msadr471 

> And tired of this whole thing, I suppose the whole internet is gonna be BLOCKED forever in Iran.

Yes. They will give you a very limited access to certain foreign websites, though. Because you are a student after all. Take it as a win.  And you didn't show me any "logs" but from what I can see, the strange thing about your configuration is that it is working on Wifi. It should not be able to work anywhere at all, because you don't even own a domain. Without a domain, you cannot even propagate DNS, much less issuing a self-signed certificate. And ssh works because it uses the IP of your VPS, directly without a domain.

**Comment by arandomgstring** [comment]
@Azadzadeh 

> whitelisted protocols is a thing...they may let SSH pass through even to a foreign IP but randomly send RST to HTTPS traffic to the same IP

Nah, I am going to bet that the first thing they are gonna limit is SSH rather than HTTPS. Why? because the most notorious VPNs use SSH. Psiphon is the first one that comes to my mind. Besides that, you can make proxies with sshutle (or without it) very easily (easier than V2ray since you need not TLS certificate, or complex configuration) and moreover, why would they block https, the traffic of normal websites that most users need, and let ssh go through? @Msadr471 doesn't own a domain, which is why his domain doesn't resolve to any IP address. 

@Msadr471 

Are you using V2rayNG, or something on your smartphone? What application do you use to connect to your server? You can always find logs somewhere withing the application that you are using to connect to proxy server. For example, in the bottom of V2rayN a log like this is shown

![Capture - Copy](https://user-images.githubusercontent.com/117191003/209321926-776de12e-7f15-4e83-be86-482200a50fea.JPG)

The logs above show that my proxy is working fine. Yours is probably saying that no IP address could be found for hostname. Now this is a suggestion, but if you are OK with ssh, why don't you use it? There is absolutely no need for v2ray anyway. 

If you want to use SSH as a proxy, the only thing you need to do is **SSH tunneling**. Search it on internet, there are many resources for it. Either you run a command to make it, or you can use putty for it, doesn't matter. On Windows, you can use proxifier to tunnel all applications with ssh tunnel. On android, you can use http injector, or other applications that support ssh proxy. Same goes for IPhone.

**Comment by arandomgstring** [comment]
@Msadr471 
> so how is this one working? it doesn't change my IP?! for example it won't work on Instagram and WhatsApp or telegram! but it works on Pinterest (it's blocked too).

It's easy to answer this question. First of all, you need to know that many websites, have many IPs, not only one. For example, there should be at least 1000 IPs for Youtube, I guess. Finding every single IP and blocking it is a pain. So censor will block a site according to its domain. For example, everytime that you type youtube, a dns request is sent from your browser to ISP, asking what is IP address of youtube? ISP replies with a fake IP address, so you cannot open youtube, even though the real IP addresses of youtube is not blocked. 

If you manage to find all youtube's IP for every single domain inside youtube (that you cannot see unless you use wireshark) you will be able to open youtube without a vpn. Some times though, some apps such as telegram and whatsapp use a few IPs. So censor can easily block them directly with their IP address. It has nothing to do with DNS (you are not asking what is IP address of telegram, your telegram application knows it) it's a direct block on IP itself, you cannot bypass it without a VPN. 

As for difference between ISPs, well their devices are different, their traffic are different, etc. For example a smaller ISP need to let some traffic to go through, otherwise it goes bankrupt. Big companies such as Irancell don't care about these things.

**Comment by arandomgstring** [comment]
@Azadzadeh 

But he says that he cannot connect to his proxy on Mobile ISPs at all! He didn't say that he can connect but it is slow or packet loss is high or something. It's beyond throttling, there has to be simple reason such as not having domain for this type of problem. Proxy apps too won't break, they are designed in a way that they re-establish their aborted connections. At least, it is the case for V2rayN. And if you were censor, would you rather to block access of tech people (who make proxies) or normal users?

**Comment by bensafai** [comment]
> 

![result](https://user-images.githubusercontent.com/4588709/209396859-601f04e1-9b75-48d6-a285-08053ccdd103.jpg)

The numbers visible in the image relate to the setting up of v2ray that connected with an acceptable speed for instagram usage from yesterday. it's a really muddy situation. I can just say, that there is no best solution anywhere. The Law of the Jungle dictates which configuration is the best.

1: Vless+ws+tls+443+Cloudflare-CDN (Hetzner-DE)
2: Vmess+ws+tls+443+Cloudflare-CDN (Hetzner-DE)
3: Vless+tcp+tls+443+alpn(http/1.1) (Hetzner-DE)
4: Vless+tcp+xtls(rprx-direct)+443+Cloudfront-CDN (AWS-DE)
5: Vless+tcp+xtls(rprx-splice)+443+Cloudfront-CDN (AWS-DE)
6: Vless+ws+tls+443+Cloudfront-CDN (AWS-DE)
7: Vmess+ws+tls+443+Cloudflare-CDN (AWS-DE)
8: Trojan+grpc(gun)+tls+443+alpn(h2)+Cloudfront-CDN (AWS-DE)
9: Vless+grpc(multi)+tls+443+alpn(h2,http/1.1)+Cloudflare-CDN (AWS-DE)

**Comment by free-the-internet** [comment]
> > > tell us if you find out! Xray+Vless+TCP+TLS seem to be the most resilient. but it doesn't protect against the active probe (see [#166 (comment)](https://github.com/net4people/bbs/issues/166#issuecomment-1356813450)). you need to have a camouflage website on port 443 with HTTPS and redirect suspicious packets to this endpoint so the censor does not add your domain/IP to _the list_.
> > 
> > 
> > Do you have a tutorial for having fallback website for Vless + TCP + TLS? I see one here: https://henrywithu.com/coexistence-of-web-applications-and-vless-tcp-xtls/ but it looks complex, I feel it must be simpler than that. FYI, I used this: https://github.com/v2fly/v2ray-examples/tree/master/VLESS-TCP-TLS-proxy%20protocol, but I get nginx config error.
> 
> see this repo: https://github.com/reeceyng/v2ray-agent It works for me.

I used the script, but I had to do some modifications to be able to use it. The obvious one is the domain name: after typing it, the script can't get the IP address.
Also they are using `pgrep -f xray/xray` which I think it's wrong and I changed it to simple pgrep -f xray (though I knew that a lot of people using(used it and it's working for them).
But, my fake site didn't came up.
FYI, I don't want to use CDN in current situation, it can cause complications.
@arandomgstring people said you have a working fake site of 443 with Vless TLS TCP on the port 443. If it the case and it is possible for you, please give me the script of the config files.

**Comment by free-the-internet** [comment]
> > 
> 
> ![result](https://user-images.githubusercontent.com/4588709/209396859-601f04e1-9b75-48d6-a285-08053ccdd103.jpg)
> 
> The numbers visible in the image relate to the setting up of v2ray that connected with an acceptable speed for instagram usage from yesterday. it's a really muddy situation. I can just say, that there is no best solution anywhere. The Law of the Jungle dictates which configuration is the best.
> 
> 1: Vless+ws+tls+443+Cloudflare-CDN (Hetzner-DE) 2: Vmess+ws+tls+443+Cloudflare-CDN (Hetzner-DE) 3: Vless+tcp+tls+443+alpn(http/1.1) (Hetzner-DE) 4: Vless+tcp+xtls(rprx-direct)+443+Cloudfront-CDN (AWS-DE) 5: Vless+tcp+xtls(rprx-splice)+443+Cloudfront-CDN (AWS-DE) 6: Vless+ws+tls+443+Cloudfront-CDN (AWS-DE) 7: Vmess+ws+tls+443+Cloudflare-CDN (AWS-DE) 8: Trojan+grpc(gun)+tls+443+alpn(h2)+Cloudfront-CDN (AWS-DE) 9: Vless+grpc(multi)+tls+443+alpn(h2,http/1.1)+Cloudflare-CDN (AWS-DE)

Looks random to me. I would say there is no correlation between the protocol and successful connectivity, because all the setups are the same from the censors point of view. Maybe it is more related to the IP and random droppings? or settings of the users, specially the DNS thing. Also the blockage of the CDN IP.

**Comment by free-the-internet** [comment]
> Is it doable to have a Tor node or Relay on my VPS? and I can use it, in this case, I think it has a better speed, Right? tor is very good in Iran and works on every platform and ISPs. ![image](https://user-images.githubusercontent.com/49529241/209435097-9c56f882-065b-4b6e-acd7-b2c27cf2d1ec.png) both bridges are working.

Unfortunately snowflake bridges are commissioned by the broker. Only you can make and use private obfs4 bridges on your VPS. Search Tor documentation, they explained. But the tor speed is effected by the tor nodes, so you can not guarantee your speed.
FYI: 3 months ago I created a Tor obfs4 bridge, but it couldn't connect in Iran. Please try and report here. Thanks.

**Comment by Hadi-1624** [comment]
@Msadr471  When you say through the Arvan Cloud, do you mean their CDN service? or their  vps solution

**Comment by free-the-internet** [comment]
> So, I told a friend and he helped me, now my VPS works on all ISPs, including Irancell. Solution: Irancell (obviously) limited its connection to the outside of Iran so instead of trying to communicate to the world that is limited by Irancell I routed my traffic through the "Arvan Cloud" and then the Hatzener server. for now, the only thing that Irancell sees is "Arvan Cloud" not Hatzener. also now my VPS has an SSL certification and a domain. In the end, my speed now is super fast:
> 
> WiFi: ![Screenshot_2022-12-25-00-22-28-503_org zwanoo android speedtest] [image] (https://user-images.githubusercontent.com/49529241/209450997-77a7eb30-2b4e-4d51-9bbd-32ffa61949d4.jpg)
> 
> Irancell: ![Screenshot_2022-12-25-00-30-34-404_org zwanoo android speedtest] [image] (https://user-images.githubusercontent.com/49529241/209451024-eac8f092-45cb-4cc5-802f-4c6359abf41d.jpg)

I really discourage the use of any local VPS, unless there is a emergency when there is a shutdown.
IMHO, I consider this falling into so called "layered internet" (layered in terms of the population) by your own hand, as for getting a VPS ,they need your identity. So, what would happen is this: The Internet access for the ordinary people would be cut, and those who have the services like you have, could be identified.

**Comment by a764934018** [comment]
> > see this repo: https://github.com/reeceyng/v2ray-agent
> 
> oh, it's a maintained fork of [mack-a script](https://github.com/mack-a/v2ray-agent)! His script looked very robust when I tried it (trojan option). I don't know why he suddenly removed his repo, I mean, did he find a security flaw and didn't want to spend time on it (so he removed it) or what?

most likely is that the author was get caught by goverment, like the author of original python version's shadowsocks: https://github.com/shadowsocks/shadowsocks. many authors who write censorship circumvention tools or scripts were get caught before in China.

**Comment by free-the-internet** [comment]
Guys in Iran, I need checks.
Tor obfs4 bridges (of course with non-blocked IPs) are working in Iran? I had it 2 3 months ago, never worked; again I checked these days, it doesn't work with non-blocked private bridges.
@arandomgstring and others ...

**Comment by arandomgstring** [comment]
@free-the-internet 
It has been a while since obfs4 bridges has stopped working. With exception of snowflake, other bridges apparently don't work.

**Comment by free-the-internet** [comment]
> @free-the-internet It has been a while since obfs4 bridges has stopped working. With exception of snowflake, other bridges apparently don't work.

Do you think it's because of the fingerprinting? How we can use meeklite in a tor client? or build a bridge with meeklite compatibility? @wkrp Could you help us please?

I've got the reports that VLESS + TCP or Trojan + TLS is connecting by not usable at all, specially on mobile operators.

**Comment by DevilHimSelf666** [comment]
i tried with 3 domain 
1. xxx.nl
2. xxx.com
3. xxx.ir

with one server through Cloudflare the result was so weird 
with the nl domain my speed was awful (download : 1.13 Mbps  )
with .com the speed was very good (download : 72.4  Mbps  ) but unstable ( i could not watch youtube for more than 5 min ) 
with .ir the speed was ok (download : 21.4  Mbps  ) but it was stable and i could watch youtube without stop

now im going to change my dns to Arvan Cloud i will let you know if there was any significant change

**Comment by wkrp** [comment]
> > It has been a while since obfs4 bridges has stopped working. With exception of snowflake, other bridges apparently don't work.
> 
> Do you think it's because of the fingerprinting? How we can use meeklite in a tor client? or build a bridge with meeklite compatibility? @wkrp Could you help us please?

If obfs4 bridges are indeed being blocked, there could be a variety of causes. It may be enumeration of the distribution system; it could be identification of high-entropy connections; it may have to do with connection lifetime or connection patterns; it could be blocking of specific foreign IP address ranges. It's hard to say.

There's really no such thing as "meek_lite". That label was originally referred to an independent implementation of the meek protocol in obfs4proxy; the "lite" was meant to indicate that [it did not have any TLS camouflage](https://gitlab.com/yawning/obfs4/-/blob/336a71d6e4cfd2d33e9c57797828007ad74975e9/ChangeLog#L53). (The [mainline meek implementation](https://gitweb.torproject.org/pluggable-transports/meek.git/) used a [headless browser for TLS camouflage](https://www.bamsoftware.com/papers/fronting/#sec:browserextension) at the time.) Later, obfs4proxy's implementation [started using uTLS](https://gitlab.com/yawning/obfs4/-/commit/4d453dab2120082b00bf6e63ab4aaeeda6b8d8a3) for TLS camouflage, so the implementation was no longer really "lite". The mainline implementation also [gained the ability to use uTLS](https://bugs.torproject.org/tpo/anti-censorship/pluggable-transports/meek/29077), as an alternative to a headless browser. There's really no difference between meek and meek_lite.

The demand for the built-in meek-azure bridge is much larger than the supply. The bridge is self-throttled to about 5 MB/s (as you can see in the [bandwidth graph](https://metrics.torproject.org/rs.html#details/B1AA2EF6435720FB710B9F94872691A2DCD49E78)) in order to control costs, because the bridge is so expensive to operate. In comparison, one of the Snowflake bridges currently averages around 240 MB/s.

It *is* possible to use meek with a personally operated meek-server, with or without a CDN in the middle. For example, you can use this bridge line:
```
Bridge meek 192.0.2.31:3 1922840D0D66CB82EACE4327F5001430227C0127 url=https://meek.bamsoftware.com/ utls=HelloChrome_auto
```
But without a CDN, blocking resistance depends on keeping the bridge URL secret. There is nothing to prevent a censor from blocking the above server. But if you set up your own server and keep it private, you can use the meek transport protocol with your own bridges.
